src/utils/request/onUnhandledRequest.ts

```diff
@@ -1,36 +1,162 @@
+  GraphQLRequestParsedResult,
+  ParsedQueryPayload,
+} from '../../graphql'
 import { getPublicUrlFromRequest } from './getPublicUrlFromRequest'
+import { isStringEqual } from '../internal/isStringEqual'
 
 type UnhandledRequestCallback = (req: MockedRequest) => void
 
-export type OnUnhandledRequest =
+export type UnhandledRequestStrategy =
   | 'bypass'
   | 'warn'
   | 'error'
   | UnhandledRequestCallback
 
+type RestRequestHandler = RequestHandler<
+  MockedRequest<DefaultRequestBodyType, RequestParams>,
+  typeof restContext,
+  ParsedRestRequest
+>
+
+type GraphQLRequestHandler = RequestHandler<
+  GraphQLMockedRequest<any>,
+  GraphQLMockedContext,
+  GraphQLRequestParsedResult<any>
+>
+
+function groupHandlersByType(handlers: RequestHandlersList) {
+  return handlers.reduce<{
+    rest: RestRequestHandler[]
+    graphql: GraphQLRequestHandler[]
+  }>(
+    (groups, handler) => {
+      const metaInfo = handler.getMetaInfo()
+      groups[metaInfo.type].push(handler)
+      return groups
+    },
+    {
+      rest: [],
+      graphql: [],
+    },
+  )
+}
+
+type RequestHandlerSuggestionList = [number, RequestHandler][]
+type ScoreGetterFn = (request: MockedRequest, handler: RequestHandler) => number
+
+function getScoreForRestHandler(): ScoreGetterFn {
+  return (request, handler) => {
+    const { mask, method } = handler.getMetaInfo()
+
+    if (mask instanceof RegExp) {
+      return Infinity
+    }
+
+    const hasSameMethod = isStringEqual(request.method, method)
+    // Always treat a handler with the same method as a more similar one.
+    const methodScoreDelta = hasSameMethod ? 0 : 0.25
+    const requestPublicUrl = getPublicUrlFromRequest(request)
+    const score = getStringMatchScore(requestPublicUrl, mask)
+
+    return score + methodScoreDelta
+  }
+}
+
+function getScoreForGraphQLHandler(
+  parsedQuery: ParsedQueryPayload,
+): ScoreGetterFn {
+  return (_, handler) => {
+    if (typeof parsedQuery.operationName === 'undefined') {
+      return Infinity
+    }
+
+    const { operationType, operationName } = handler.getMetaInfo()
+    const hasSameOperationType = parsedQuery.operationType === operationType
+    // Always treat a handler with the same operation type as a more similar one.
+    const operationTypeScoreDelta = hasSameOperationType ? 0 : 0.25
+    const score = getStringMatchScore(parsedQuery.operationName, operationName)
+
+    return score + operationTypeScoreDelta
+  }
+}
+
+function getSuggestedHandler(
+  request: MockedRequest,
+  handlers: RequestHandlersList,
+  getScore: ScoreGetterFn,
+): RequestHandler | undefined {
+  const suggestedHandler = handlers
+    .reduce<RequestHandlerSuggestionList>((acc, handler) => {
+      const score = getScore(request, handler)
+      return acc.concat([[score, handler]])
+    }, [])
+    .sort(([leftScore], [rightScore]) => {
+      return leftScore - rightScore
+    })
+    .find(([score]) => {
+      return score <= 2
+    })
+
+  return suggestedHandler?.[1]
```

Comment:
Woops, I was thrown off guard here and wanted to suggest `suggestedHandler?.[0]`.
But then I realized that 0 is the score and 1 is the handler.
Would the following make the intention more clear?

```suggestion
  const [_, suggestedHandler] = handlers
    .reduce<RequestHandlerSuggestionList>((acc, handler) => {
      const score = getScore(request, handler)
      return acc.concat([[score, handler]])
    }, [])
    .sort(([leftScore], [rightScore]) => {
      return leftScore - rightScore
    })
    .find(([score]) => {
      return score <= 2
    })

  return suggestedHandler
```

Reply:
It would, but `.find` can return `undefined`. One cannot destructure a possibly undefined value. Thus we would need a fallback in order to implement your suggestion:

```js
const [_, a] = [].find(predicate) || []
```

I think this is a little confusing.

Reply:
Oh shoot, you're right! I usually use this for `.filter` which works on an empty array, but I hadn't thought this through for `find` which can return `undefined` as you mentioned.

Sorry for the confusion!